Export style 구현

GodDrinkTeJAVA opened the pull request:
![styledmap](https://user-images.githubusercontent.com/46532449/101281350-cc698900-3811-11eb-9f39-0c51e1e161e4.png)

아직 완성은 되지 않은 상태지만, 구현에서 참조가 되도록 우선 올립니다.

## 구현 부문
* 스타일을 JSON 형태로 추출하여 각각 원본과 URIEncoded된 형태로 제공
* 변경점이 없는 경우 결과로 나오는 JSON에서 제거
* 내보내기를 클릭하면 이를 각각 보여주는 modal 창 표시

## TODO
* 아직 하위항목이 안 보이게 하는데 조정이 필요합니다... isChange가 있는 항목도 있고 없는 항목도 있어서...
* 레이아웃을 조정하는 과정중입니다.

GodDrinkTeJAVA committed:
[fix] 가시성 설정시 배경이 작동하지 않는 오류 수정

background인 경우 오직 color만을 이용해서 설정하기에 생긴 문제였습니다.
stylekeytype에 따라 적절히 나눠지도록 설정하여 문제를 해결했습니다.

GodDrinkTeJAVA committed:
[refactor] 이전 커밋에서 설정한 조건문 리팩토링

1. 생각해보니 visibility는 배경을 따로 처리하지 않아도 됐습니다.
- 그냥 조건을 추가하여 visibility라면 따로 처리하지 않는 것으로 바꾸었습니다.
2. literal로 주어진 조건문의 일부를 enum으로 처리하였습니다.
- 이제 보다 관리하기 수월해질 것이라 기대합니다.

GodDrinkTeJAVA committed:
[fix] DetailTypeSubList의 key값 지정

지정하지 않아서 리액트가 내색을 하던 현상을 수정하였습니다

GodDrinkTeJAVA committed:
[feat] 정보 추출을 위한 리듀서 초기화

action, reducer 등을 정의했습니다.
ActionType도 action type GET을 추가해줬습니다

GodDrinkTeJAVA committed:
[feat] 기본적인 형태의 stringify 구현

1. 모달을 위한 컴포넌트 구현
- 모달 자체를 위한 ExportModal을 초기화하였으며,
  이를 SidebarFooter에 삽입했습니다.
2. 이를 위한 React hook 변경
- useExportStyle을 통해 redux로 현재 상태를 가져온 뒤 stringify를 하고,
  useSidebarFooter를 통해 값을 전달해주고 있습니다.
  이후 받아온 객체를 정제하는 과정을 거칠 예정입니다.
- 이 구조가 맞는지 잘 모르겠습니다... 이후에 변경하겠습니다...
3. 기존의 linter 에러 수정
- action에서 객체와 세미콜론 삽입을 진행했습니다.

GodDrinkTeJAVA committed:
[feat] json을 표시하기 위한 컴포넌트 구현

전반적인 디자인적 일관성을 위해서 ImportModal의 구조를 따랐습니다.
이를 위해서 기존에 ImportModal에 있는 컴포넌트를 common.tsx로 옮겼습니다.
현재 모달 둘이 Import와 Export가 따로 존재하는데,
이를 합치는 방식으로 리팩토링이 가능하지 않을까...생각이 듭니다.

GodDrinkTeJAVA committed:
[feat] feature, subfeature를 위한 필터링 구현

각 단계마다 각 객체의 entries를 통해 순회를 한 뒤 조건에 따라 뺍니다.
예를 들어 feature의 경우 길이가 0인 feature를 제외하고 있으며
subfeature는 마찬가지로 길이가 0인 subfeature,
그리고 key 중 의미가 없는 all을 필터링하는 식입니다.
이렇다보니 src\store\common\type.ts의 각 타입을 사용하고자 했는데
필터링하면 없어도 되는거니깐 타입스크립트님을 맞출 수가 없더라고요.
결국 useExportStyle 내에 다 ?:를 이용해서 같은 이름의 타입을 정의했는데
이게 맞는지 나중의 논의하고 싶습니다 ;3

GodDrinkTeJAVA committed:
[feat] 추가적인 필터링 구현 및 초기화, 컴포넌트 일부 수정

element와 subElement를 위한 필터를 구현했습니다.
현재 null인 속성이 나오고 있는 문제가 있습니다. 추후 수정하겠습니다.
또한, json으로 나온 값이 적절하게 출력되기 위해서
pre 태그를 추가하였습니다.

GodDrinkTeJAVA committed:
[fix] footer로 보내는 데이터를 원본으로 보내는 것으로 변경

이전에는 JSON stringify를 hook 단계에서 해서 보내줬지만,
URL과 JSON 형태 양쪽을 제공하기 위해서 원본 데이터를 전달해주기로 했습니다.

GodDrinkTeJAVA committed:
[feat] URI export 추가

별 건 없고 레이아웃에서 JSON을 일렬로 뽑은 다음에 uriencode 했습니다
쿼리 키는 나중에 합칠 때 확정하도록 하겠습니다

qkrdmstlr3 commented on src/store/style/getReducer.ts:
```diff
@@ -38,6 +38,9 @@ export default function getReducer(IDX: number): ReducerType {
         return initialState;
+      case GET:
+        console.log(initialState);
+        return initialState;
```
GET의 목적은 무엇인가요?? 단순히 가져오는 것이라면 useSelector로 가져올 수는 없을까요??

Eunsol0410 replied:
저도 이부분이 궁금합니다!

GodDrinkTeJAVA replied:
앗...아앗... 예전에 api fetch하던 습관 때문에 만들어놓고서는 안 쓰고 있습니다.
곧바로 삭제하도록 하겠습니다!!!

qkrdmstlr3 commented on src/hooks/sidebar/useExportStyle.ts:
```diff
@@ -0,0 +1,131 @@
+  section?: SubElementType | null;
+  labelText?: SubElementType | null;
+  labelIcon?: StyleType | null;
+}
```
기존에 선언한 타입들로는 해결이 안되셔서 새로 선언하신 것인가요?

GodDrinkTeJAVA replied:
네 맞습니다. common/type에서 정의한 거는 undefined가 없어서 json으로 없는 부분을 생략할 수가 없더라고요.

qkrdmstlr3 commented on src/hooks/sidebar/useExportStyle.ts:
```diff
@@ -0,0 +1,131 @@
+
+function filterStyle(style: StoreDataType): StoreDataType {
+  const ret = Object.entries(style).reduce((accu, [key, subFeature]) => {
+    const filteredValue = filterSubFeature(subFeature as SubFeatureType);
```
accu대신 좀 더 명시적으로 이름을 정해주는 것은 어떨까요?

qkrdmstlr3 reviewed (commented):
reduce들의 생김새가 같은 듯 조금씩 다르네요... 수고하셨습니다!

Eunsol0410 commented on src/components/Sidebar/SidebarContentMore/DetailTypeSubList.tsx:
```diff
@@ -59,11 +59,11 @@ function DetailTypeSubList({
     return child.childrenProps.length !== 0 ? (
-      <div>
+      <div key={child.title}>
         <Text padding="second">{child.title}</Text>
```
key를 추가하셨군요! 👍

Eunsol0410 commented on src/hooks/sidebar/useExportStyle.ts:
```diff
@@ -0,0 +1,131 @@
+        return accu;
+      }
+
+      const { isChange, ...changedStyleKey } = styleKey;
```
isChanged 오타인 것 같습니다!

Eunsol0410 replied:
그런데 SubElementType에는 isChanged가 없지 않나요?

GodDrinkTeJAVA replied:
우선 오타는 현재 로컬 저장소에서는 고친 상태입니다!
그런데 isChanged가 labelIcon 쪽에 있지 않던가요? 우선 확인해보겠씁니다. 아니면... 제가 타입 밀려쓰기를 하고 있는걸지도 모르겠네요 (   ..)

GodDrinkTeJAVA replied:
우선 오타는 알고 보니 아직도 고치지 않은 상태여서 수정하였습니다
그런데 isChanged는 subElement에 있지 않던가요? labelIcon과 같이 fill, stroke 구분이 없는 요소는 있던 것으로 기억하는데...

Eunsol0410 replied:
아... 그런 의미였군요!!! subElement는 fill, stoke에 해당하고 labelIcon은 바로 Style을 정의한다고 생각하고 있어서 그랬습니다.  
그렇다면 해당 예외체리를 밖에서 할 수 있는 방법이 있을까요?!?

Eunsol0410 commented on src/hooks/sidebar/useExportStyle.ts:
```diff
@@ -0,0 +1,131 @@
+    return {};
+  }
+
+  const { isChanged, ...changedElement } = element;
```
element에도 상태상에 isChanged는 없었던 것 같습니다. 필요해서 추가를 하셨나요?

Eunsol0410 commented on src/hooks/sidebar/useExportStyle.ts:
```diff
@@ -0,0 +1,131 @@
+  };
+
+  return { exportStyle };
+}
```
useExportStyle 함수가 연정님이 구현하는 부분에서도 사용할 수 있을 것 같은데, 그렇다면 따로 분리해도 좋을 것 같아요!

Eunsol0410 reviewed (commented):
수고하셨습니다!!! 준희님 👍 끝까지 함께 해주셔서 든든하네요

GodDrinkTeJAVA committed:
[fix] element 단계에서 비정상적으로 출력되는 json 일부 수정

기껏 isChanged를 거르고서 걸러낸 객체를 넣지 않고 있었습니다.
쓸데없이 시간을 많이 잡아먹었네요 죄송합니다 흐극

GodDrinkTeJAVA committed:
[refactor] 사용되지 않는 action 삭제

예전에 redux를 썼을 때 거의 웹 통신을 할 때 썼던지라 넣었던건데...
지금은 의미 없는 선택지가 됐네요. 관련 구문을 제거했습니다.

GodDrinkTeJAVA committed:
[fix] isChanged가 표시되지 않는 항목이 나오는 문제 수정

reduce를 잘못 설정해 각 항목 안으로 상위 항목이 들어가서 생긴 문제였습니다.

GodDrinkTeJAVA committed:
[feat] 스타일 leaf property에서 변경되는 항목만 표시되는 기능 구현

이전까지 isChanged를 통해서 비교를 한 뒤,
속성 내에서는 지금까지의 속성의 key값을 모은 currentLocation을 통해
기본 style을 구하고, 이를 기반으로 비교해서 변한 속성만 남깁니다.
더불어, leaf style property에서 isChanged를 제거하는 기능 또한
이 함수로 위치를 변경하였습니다.

GodDrinkTeJAVA commented:
위의 commit으로 element 이하 항목이 이제 제대로 표시됩니다.

GodDrinkTeJAVA committed:
[chore] console.log 제거 및 미사용 변수 제거

GodDrinkTeJAVA commented:
위의 commit으로 element 이하 항목이 이제 제대로 표시됩니다.

gitgitWi commented on src/hooks/sidebar/useExportStyle.ts:
```diff
@@ -0,0 +1,183 @@
+  }
+
+  const ret = Object.entries(subElement).reduce((accu, [key, styleKey]) => {
+    if (!styleKey.isChanged) {
+      return accu;
+    }
+
+    return styleKey.isChanged
+      ? {
```
이부분 코드 `styleKey.isChanged` 의 T/F 값으로 결정되는데, 하나로 합칠수 있지 않을까요?

gitgitWi reviewed (commented):
수고하셨습니다!
리뷰어 지정 부탁드려요 ㅋㅋㅋㅋ
